py-rattler/rattler/version/with_source.py

```diff
@@ -0,0 +1,447 @@
+from __future__ import annotations
+
+from typing import List, Optional, Tuple, Union
+
+from rattler.version import Version
+
+from rattler.rattler import PyVersionWithSource, InvalidVersionError
+
+
+class VersionWithSource:
+    """
+    Holds a version and the string it was created from. This is useful if
+    you want to retain the original string the version was created from.
+    This might be useful in cases where you have multiple strings that
+    are represented by the same [`Version`] but you still want to be able to
+    distinguish them.
+
+    The string `1.0` and `1.01` represent the same version. When you print
+    the parsed version though it will come out as `1.0`. You loose the
+    original representation. This struct stores the original source string.
+    """
+
+    def __init__(self, source: str, version: Optional[Version] = None):
+        if isinstance(source, str):
+            if version is None:
+                version = Version(source)
+
+            if isinstance(version, Version):
+                # maybe we should use _inner for inner FFI objects everywhere?
+                self._inner = PyVersionWithSource(version._version, source)
+            else:
+                raise TypeError(
+                    "VersionWithSource constructor received unsupported type "
+                    f" {type(version).__name__!r} for the `version` parameter"
+                )
+        else:
+            raise TypeError(
+                "VersionWithSource constructor received unsupported type "
+                f" {type(version).__name__!r} for the `source` parameter"
+            )
+
+    @property
+    def version(self) -> Version:
+        """
+        Returns the `Version` from current object.
+
+        Examples
+        --------
+        >>> v = VersionWithSource("1.0.0")
+        >>> v.version
+        Version("1.0.0")
+        >>> v2 = VersionWithSource("1.0.0", v.version)
+        >>> v2.version
+        Version("1.0.0")
+        """
+        return Version._from_py_version(self._inner.version())
+
+    @property
+    def epoch(self) -> Optional[int]:
+        """
+        Gets the epoch of the version or `None` if the epoch was not defined.
+
+        Examples
+        --------
+        >>> v = VersionWithSource('2!1.0')
+        >>> v.epoch
+        2
+        >>> v2 = VersionWithSource('2!1.0', v.version)
+        >>> v2.epoch
+        2
+        """
+        return self._inner.epoch()
+
+    def bump(self) -> VersionWithSource:
```

Comment:
I think this should return a Version instead. Since you modify the original version anyway it doesnt make sense to keep the source.